ssl/ssl_key_share.cc

```diff
@@ -294,6 +295,233 @@ class CECPQ2KeyShare : public SSLKeyShare {
   HRSS_private_key hrss_private_key_;
 };
 
+class KEMKeyShare : public SSLKeyShare {
+ public:
+  KEMKeyShare(int nid, uint16_t group_id) : nid_(nid), group_id_(group_id) {}
+
+  uint16_t GroupID() const override { return group_id_; }
+
+  bool Offer(CBB *out) override {
+    // Ensure |out| is valid, has no children, and has been initialized.
+    // We check that |out| has no children because otherwise CBB_data()
+    // will produce a fatal error by way of assert(cbb->child == NULL).
+    if (!out || out->child || !CBB_data(out)) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_PASSED_NULL_PARAMETER);
+      return false;
+    }
+
+    // ctx_ should not have been initialized (by e.g. a previous call
+    // to Offer()).
+    if (ctx_) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_SHOULD_NOT_HAVE_BEEN_CALLED);
+      return false;
+    }
+
+    // Initialize the KEM ctx
+    ctx_.reset(EVP_PKEY_CTX_new_id(EVP_PKEY_KEM, nullptr));
+    if (!ctx_ || !EVP_PKEY_CTX_kem_set_params(ctx_.get(), nid_)) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_INTERNAL_ERROR);
+      return false;
+    }
+
+    // Generate a KEM keypair and retrieve the length of the public key
+    EVP_PKEY *raw_key = nullptr;
+    if (!EVP_PKEY_keygen_init(ctx_.get()) ||
+        !EVP_PKEY_keygen(ctx_.get(), &raw_key) ||
+        !raw_key ||
+        !EVP_PKEY_get_raw_public_key(raw_key, nullptr, &public_key_len_)) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_INTERNAL_ERROR);
+      return false;
+    }
+
+    // Retain the private key for decapsulation
+    UniquePtr<EVP_PKEY> pkey(raw_key);
+    ctx_.reset(EVP_PKEY_CTX_new(pkey.get(), nullptr));
+    if (!ctx_) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_INTERNAL_ERROR);
+      return false;
+    }
+
+    // Write the public key to |out|
+    Array<uint8_t> public_key;
+    if (!public_key.Init(public_key_len_)) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_MALLOC_FAILURE);
+      return false;
+    }
+
+    size_t public_key_bytes_written = public_key_len_;
+    if (!EVP_PKEY_get_raw_public_key(pkey.get(), public_key.data(),
+                                     &public_key_bytes_written) ||
+                                     public_key_bytes_written != public_key_len_ ||
+                                     !CBB_add_bytes(out, public_key.data(), public_key_len_)) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_INTERNAL_ERROR);
+      return false;
+    }
+
+    return true;
+  }
+
+  bool Accept(CBB *out_public_key, Array<uint8_t> *out_secret,
```

Comment:
`out_public_key`? shouldn't this be `out_ciphertext` or something like that?

Reply:
Like above, I'm following the established convention set by CECPQ2:

https://github.com/aws/aws-lc/blob/dbb1ea4499804e046276a1d1ab61187e142d3ff4/ssl/ssl_key_share.cc#L229-L230

https://github.com/aws/aws-lc/blob/dbb1ea4499804e046276a1d1ab61187e142d3ff4/ssl/ssl_key_share.cc#L267-L268

 and the function definition:

https://github.com/aws/aws-lc/blob/dbb1ea4499804e046276a1d1ab61187e142d3ff4/ssl/internal.h#L1105-L1120